### PR TITLE
Retry ISM save upon conflicting IS error

### PR DESCRIPTION
**Pull request by soltysh**
@ncdc this addressed the problem I'm having in #4853 regarding creating IS and immediately posting ISM right after that, which might conflict with import logic. ptal.

### CODE REVIEW

**Comment by soltysh** [comment]
@miminar this is solution to the problem we've talked earlier today.

**Comment by ncdc** [comment]
Also cc @smarterclayton

**Comment by smarterclayton** [comment]
I'm going to put a real retry function in upstream for use later.

File: pkg/dockerregistry/server/repositorymiddleware.go [hunk 1]
```diff
@@ -225,6 +229,28 @@ func (r *repository) Put(ctx context.Context, manifest *manifest.SignedManifest)
 	return nil
 }
 
+// createImageStreamMapping creates ISM, the operation is retried if conflict at ImageStream
+// level is returned as a result, in all other cases the actual error is returned.
+func (r *repository) createImageStreamMapping(namespace string, ism *imageapi.ImageStreamMapping, retryCount int) error {
+	if retryCount == ismRetries {
+		return fmt.Errorf("reched ismRetries limit")
```
Comment (miminar) [inline pkg/dockerregistry/server/repositorymiddleware.go]: `s/reched/reached/`

File: pkg/dockerregistry/server/repositorymiddleware.go [hunk 2]
```diff
@@ -225,6 +229,28 @@ func (r *repository) Put(ctx context.Context, manifest *manifest.SignedManifest)
...
+		statusErr, ok := err.(*kerrors.StatusError)
+		if !ok {
+			return err
+		}
+
+		status := statusErr.ErrStatus
+		if status.Code != http.StatusConflict || status.Details.Kind != "imageStream" || status.Details.Name != ism.Name {
+			return err
+		}
+		return r.createImageStreamMapping(namespace, ism, retryCount+1)
```
Comment (miminar) [inline pkg/dockerregistry/server/repositorymiddleware.go]: I find the incrementation of retry counter confusing. When a func is invoked with `retrycount == 0`, I'd expect it to fail without a retry. I'd rather follow our pattern in various controllers where the counter is decremented until it reaches 0.

Comment (soltysh) [inline pkg/dockerregistry/server/repositorymiddleware.go]: I think that'll be replaced with the logic @smarterclayton created upstream, I'll cherry-pick his PR and update this one.

**Comment by soltysh** [comment]
@miminar addressed one of your comment, the other was replaced with @smarterclayton upstream PR. ptal once again

**Comment by miminar** [comment]
LGTM

**Comment by soltysh** [comment]
Rebased. @smarterclayton I don't see any movement in the upstream PR, is it going to be as is, or are there some changes expected there? Iow. is it ok to merge this?

**Comment by smarterclayton** [comment]
[merge]

**Comment by smarterclayton** [comment]
Actually, use DefaultRetry, not DefaultBackoff

**Comment by soltysh** [comment]
@smarterclayton done

**Comment by smarterclayton** [comment]
[merge]

**Comment by soltysh** [comment]
Either I'm very lucky with this error in tests or else...
```
Error pushing to registry: Put http://172.30.63.40:5000/v2/cache/ruby-20-centos7/blobs/uploads/362468b4-4b8c-4c9a-ae14-5c1560aade6a?_state=QDenasV8m_qM3UZ9NFKczjCWdHZ0hlEYeXoA-aU31Jp7Ik5hbWUiOiJjYWNoZS9ydWJ5LTIwLWNlbnRvczciLCJVVUlEIjoiMzYyNDY4YjQtNGI4Yy00YzlhLWFlMTQtNWMxNTYwYWFkZTZhIiwiT2Zmc2V0IjowLCJTdGFydGVkQXQiOiIyMDE1LTEwLTIzVDE3OjU4OjM2LjcxMzA3NDA1NFoifQ%3D%3D&digest=sha256%3Afda2ed1242ded4612516a67d6405b847e84cb9eeb98a2623d24f7bcd8d66c171: dial tcp 172.30.63.40:5000: network is unreachable
```
Let me retry, if that fails again, I'll propose returning to DefaultBackof.
[merge]

**Pull request merged**